カレンダーを表示する機能を備えたコマンドラインプログラムを実装

tuna24365 opened the pull request:
# 実装内容
コマンドラインからオプションで年月を指定してカレンダーを表示できるプログラムを実装しました。

## 主な機能
- コマンドライン引数 (-y、-m) を`minimist`により解析し、ユーザーが指定した年月のカレンダーを表示します。
- 引数を指定しない場合、自動的に現在の年月のカレンダーが表示されます。

## 注記
- 例外処理を用いたエラーハンドリングの実装は組み入れていません。
- 基本的なカレンダー表示のみのシンプルなプログラムです。

## テスト
✅ `ESLint`と`Prettier`を通過しています。
✅ 異なる年月をコマンドライン引数として渡し、カレンダーが正確に表示される事を確認しました。
✅ `1970年1月`と`2100年12月`のカレンダーが正しく表示される事を確認しました。

tuna24365 committed:
Add command line program for displaying calendar

JunichiIto commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,82 @@
+const HEADER_WIDTH = 14;
+const DAY_WIDTH = 2;
+const BLANK_DAY = " ".repeat(3);
+const FIRST_DAY_OF_MONTH = 1;
```
これはマジックナンバーではないので定数化しなくてもいいのでは、と思いました。

tuna24365 replied:
コメントありがとうございます🙏✨
今回はコードの`可読性`と`保守性`を意識して定数化しました。
しかし、ご指摘のとおりで例えば`FIRST_DAY_OF_MONTH（月初めの日）`を表す`1`は自明で`マジックナンバー`ではないことや、他の定数も`再利用`していないことを考えると少々冗長で、あらためて定数化の必要性は薄いと感じました💦

ご意見を受け、定数の使用を見直そうと思います。

JunichiIto commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,82 @@
+};
+
+const isSaturday = (year, month, day) => {
+  return new Date(year, month - 1, day).getDay() === 6;
```
このyearとmonthは引数で渡された値ですが、それ以外のメソッドではグローバル変数（？）のyearとmonthに依存しています。
こういう混乱の原因になるので、9〜26行目の処理はrunメソッドに入れて、グローバル変数をなくしたいと思いました。
（寿命の長い（＝scopeの広い）変数はなるべくなくした方がコードの見通しが良くなります）

tuna24365 replied:
コメントありがとうございます🙏
> こういう混乱の原因になるので、9〜26行目の処理はrunメソッドに入れて、グローバル変数をなくしたいと思いました。
>（寿命の長い（＝scopeの広い）変数はなるべくなくした方がコードの見通しが良くなります）

グローバル変数を無くしてrunメソッド内に収めるというアイデア、大変参考になります😀✨
ご指摘いただいた内容で修正しようと思います！

JunichiIto commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,82 @@
+  for (let day = FIRST_DAY_OF_MONTH; day <= dayInMonth; day++) {
+    monthDates.push(generateDayFormat(day));
+    if (isSaturday(year, month, day)) {
+      monthDates.push("\n");
```
この`push("\n");`をなくして、79行目の`join("\n");`でひとまとめにした方が、コードとしてはきれいだと思います。
（今は同じ役割の処理を2箇所でやっている状態）

tuna24365 replied:
コメントありがとうございます🙏✨
68行目`push("\n");`でmonthDatesの日付の曜日から週末を判定し、改行コードを挿入する処理をしているのでそのまま削除するのは難しいと考えています😥

理解不足で申し訳ありませんが、ひとまとめにする方法についてもう少し詳しく教えていただけますでしょうか🙇‍♂️

JunichiIto replied:
最後のrunで、今のロジックだと

```ruby
irb(main):001* [
irb(main):002*   "1",
irb(main):003*   "2",
irb(main):004*   ["3", "4"].join("\n")
irb(main):005> ].join("\n")
=> "1\n2\n3\n4"
```

とやっていますが、これを

```ruby
irb(main):017* [
irb(main):018*   "1",
irb(main):019*   "2",
irb(main):020*   "3",
irb(main):021*   "4"
irb(main):022> ].join("\n")
=> "1\n2\n3\n4"
```

となるように変えるといいかも？

（Rubyで書きましたがJSでも考え方は同じです〜）

tuna24365 replied:
わかりやすいヒントを有り難うございます🙏✨

示していただいた実行例を読み解くと、配列はrunメソッドのreturn文内のもので、その中でも`["3", "4"]`の配列generateMonthDatesメソッドの返り値に相当するものですね💡

私の書いた処理ではgenerateMonthDatesメソッド内でmonthDates配列を作って日付を格納し、`.join`で結合後さらにrunメソッド内で`.join`で結合しています💦
この点を`（同じ役割の処理を2箇所でやっている状態）`と仰っているのだと理解しました😄

ヒントを私のコードに当てはめると、generateMonthDates内では配列を使用することを止めて直接文字列を結合し、仕上げにrunメソッドの`.join`でひとまとめにするという事で改善できそうです🎵

JunichiIto reviewed (commented):
コメントしました〜。

tuna24365 committed:
Improved calendar code structure and simplicity.
- Removed global variable declarations from calendar code.
- Moved global variables into the run method.
- Simplified generateMonthDates method to use string concatenation and single .join in run method.

tuna24365 commented:
ご指摘いただいた内容を全て修正したコミットになります。
ご確認をお願いいたします🙇‍♂️

なお、以下の動作が問題ない事を確認しました。
✅ ESLintとPrettierをパスしています。
✅ 引数なしで実行した場合に、本日時点（2023年11月）の年月でカレンダーが表示される事を確認しました。
✅ `-y` `-m`の引数をそれぞれ単独指定した場合でもカレンダーが正しく表示される事を確認しました。
✅ 1970年1月と2100年12月のカレンダーが正しく表示される事を確認しました。

JunichiIto commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,76 @@
+  const monthName = getFirstDateOfMonth(year, month).toLocaleString("en-US", {
+    month: "short",
+  });
+  return `${monthName} ${year}`.padStart(14, " ");
```
14はマジックナンバーなので定数化した方がいいかもしれませんね

tuna24365 replied:
おっしゃる通りです！
ここの定数を前回のコミットで削除してしまったので元に戻します〜

JunichiIto commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,76 @@
+  for (let day = 1; day <= dayInMonth; day++) {
+    monthDates += generateDayFormat(day);
+    if (isSaturday(year, month, day)) {
+      monthDates += "\n";
```
うーん、`join("\n")` と形が変わっただけで、やってることは結局同じですね。。
言い方を変えると、`"\n"`を登場させるのを1回だけにしたい、という感じです。
たとえば、改行文字を`\n`から`\r\n`に変更したい、となった場合に、変更する箇所が1箇所だけで済むのが理想的です。

tuna24365 replied:
新たなヒントありがとうございます🙏
認識が違っていたようです💦
他の修正を進めながら、こちらはじっくり考えようと思います。
（考え中）

tuna24365 replied:
> たとえば、改行文字を\nから\r\nに変更したい、となった場合に、変更する箇所が1箇所だけで済むのが理想的です。

想定されているケース（改行文字を変える）の意図は、
OS環境によって特殊文字（改行コード）を切り替える必要が出るためですね💡
- 改行コード
  - Windows OS：`\r\n`（CRLF）
  - Mac OS：`\n`（LF）

いただいたコメントの要件を満たす事だけで考えると、
以下のように改行コードを定数化することで、1箇所で改行コードの変更ができますが、
（そんな単純な話ではない😅？）

```javascript
const NEWLINE = "\n"; // ここで改行コードを変更（`\n`→`\r\n`）

monthDates += NEWLINE;

  ].join(NEWLINE);
```

JunichiIto replied:
> 以下のように改行コードを定数化することで、1箇所で改行コードの変更ができますが、
（そんな単純な話ではない😅？）

まあ、そりゃそうだし、とんちクイズをやるとしたら「お見事、一本！」でもいいかもしれまえんが、僕のイメージとはちょっと違いますね。。

僕のコメントの意図としてはjoinを使うなら、 **各行** のデータをまとめてくっつけるのが自然じゃないかなーと思うわけです。

```
irb(main):011* puts [
irb(main):012*   "花の色は",
irb(main):013*   "うつりにけりな",
irb(main):014*   "いたづらに",
irb(main):015*   "わが身世にふる",
irb(main):016*   "ながめせしまに"
irb(main):017> ].join("\n")
花の色は
うつりにけりな
いたづらに
わが身世にふる
ながめせしまに
```

tunaさんのロジックだと、

```
irb(main):020* puts [
irb(main):021*   "花の色は",
irb(main):022*   "うつりにけりな",
irb(main):023*   "いたづらに\nわが身世にふる\nながめせしまに"
irb(main):024> ].join("\n")
花の色は
うつりにけりな
いたづらに
わが身世にふる
ながめせしまに
```

みたいになっていて、配列の要素として「1行のデータ」と「複数行のデータ（すでにjoin済みのデータ）」が混在していて、それをjoinしてるのが少し違和感ありです。

ただ、どっちのロジックでも同じ結果が得られますし、僕のロジックが絶対的に正しいというわけではない（tunaさんのやり方でも「いいんじゃない」っていう人もいるはず）ので、何が何でも修正させたいわけではないのですが、とはいえ、ひとつの考え方として、行単位のデータを集めてきて、最後にjoinする、という方法も考えてみてほしいな〜と思っております。

tuna24365 replied:
アドバイスありがとうございます🙏✨
JunichiItoさんのおっしゃる通りだと思います。
ご指摘を受けて`join("\n");`を無闇に使うことは良くないな〜と思いつつ、
とんちのような回答になってしまいました💦

今回のコミットは以下を実現できたと思います。
> 行単位のデータを集めてきて、最後にjoinする

具体的には、
generateMonthDates内に`weeks`配列を定義し、`monthDates`要素を１週ずつ入れる処理（改行コードは入れない）に変更しました。
runメソッドのreturnではgenerateMonthDatesの返り値の配列をスプレッド構文で展開し、他関数の配列と共に`.join("\n");`で一本化してみました。

ご確認をお願いいたします🙇

JunichiIto commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,76 @@
+  return monthDates;
+};
+
+const run = () => {
```
関数の並び順ですが、個人的には以下のQiita記事のような順番の方が読みやすいと思っています。
https://qiita.com/jnchito/items/860ffa0391ade46aabf9

tuna24365 replied:
提示いただいた記事大変参考になります🙏✨
コードの読み易さの意識を欠いていました💦

JunichiIto reviewed (commented):
コメントしました！

tuna24365 committed:
Refactor calendar script for enhanced readability and add HEADER_WIDTH constant.

tuna24365 committed:
Fix formatting issues.

tuna24365 commented:
チェックを付けたものを修正いたしました。
- [x] マジックナンバー(14)の定数化
- [x] 関数の並び順を変更`（呼ぶ側を手前に、呼ばれる側を後ろ）`
- [ ] `.join`の処理を一本化する

以下の動作が問題ない事を確認しました。
- [x] `ESLint`と`Prettier`をパスしています。
- [x] 引数なしで実行した場合に、本日時点（`2023年11月`）の年月でカレンダーが表示される事を確認しました。
- [x] `-y` `-m`の引数をそれぞれ単独指定した場合でもカレンダーが正しく表示される事を確認しました。
- [x] 1970年1月と2100年12月のカレンダーが正しく表示される事を確認しました。

※ 修正直後のコミットはprettierの実行を忘れてコミットしてしまったバージョンです。
　その後に続くコミットは、しっかりフォーマット整形を行い`Prettier`と`ESlint`をパスしたバージョンになります。
（ややこしくなり申し訳ありません💦）

tuna24365 committed:
Refactor Consolidate join to run method from generateMonthDates for better efficiency.

JunichiIto commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,81 @@
+};
+
+const generateWeekHeader = () => {
+  return "Su Mo Tu We Th Fr Sa";
```
メソッド化する必要はないかもしれません

https://bootcamp.fjord.jp/pages/when-you-should-define-method

tuna24365 replied:
アドバイスありがとうございます🙏✨
ご指摘の箇所は記事中の以下の事項に該当していました💦
> #### メソッド化しない方がいいケース
> - そのメソッドが呼ばれる場所が1箇所しかない（DRYに貢献しない）
> - 定義したメソッドが大した仕事をしていない（他のメソッドを呼ぶだけ等）
> #### 大した仕事をしていないメソッドの例

`単一責任の原則`を意識し過ぎて、
あまり意味のないメソッドを量産し、無駄の多いコードになっていると自覚しました😅
`generateWeekHeaderメソッド`を無くします。

`runメソッド`のreturnがカレンダー出力の全体を担っているのでこちらに直接書くことを検討してみます。

```javascript
  return [
    generateCalendarHeader(year, month),
    "Su Mo Tu We Th Fr Sa",
    ...generateMonthDates(year, month),
  ].join("\n");
```

JunichiIto commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,81 @@
+
+  for (let day = 1; day <= dayInMonth; day++) {
+    monthDates += generateDayFormat(day);
+    if (isSaturday(year, month, day) || day == dayInMonth) {
```
```suggestion
    if (isSaturday(year, month, day) || day === dayInMonth) {
```

にしましょう

tuna24365 replied:
ありがとうございます！
`厳密等価演算子`の方が適切です💦
こちらも修正します〜

JunichiIto commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,81 @@
+};
+
+const generateMonthDates = (year, month) => {
+  const dayInMonth = getLastDateOfMonth(year, month).getDate();
```
```suggestion
  const lastDay = getLastDayOfMonth(year, month);
```

みたいな感じにした方がシンプルでいいかもしれません。

tuna24365 replied:
そちらの変数名の方がわかりやすいです✨
修正します〜

JunichiIto commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,81 @@
+
+const generateDayFormat = (day) => {
+  const dayString = day.toString().padStart(2, " ");
+  return dayString + " ";
```
土曜日に無駄な空白がうしろにくっついちゃうので、それをなくせたらなくしたいですね

tuna24365 replied:
ありがとうございます😄（そこまで配慮が至りませんでした💦）

最初は`monthDates`の週末に空白を入れないよう判定を入れるべきか悩みましたが、少々手間が掛かりそうだったのでMDNを調べてみました。
`trimEnd`を使えば、一度出来てしまった空白を後から削除できそうなので週末判定`isSaturday`の中で使ってみました。

https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/String/trimEnd

⚪️修正前
<img width="1066" alt="image" src="https://github.com/tuna24365/js-practices/assets/76990362/73338317-a47b-4492-92bf-9563aa5108fc">

⚪️修正後（trimEnd）
<img width="1074" alt="image" src="https://github.com/tuna24365/js-practices/assets/76990362/41e0dd3c-f756-4b37-a683-1fa9a4644b9c">

空白が取り除けましたので、こちらでコミットしてみます〜

JunichiIto reviewed (commented):
コメントしました〜。

tuna24365 committed:
Refactor: Strict Equality, Simplified Week Header, and TrimEnd Whitespace.

- Refactor to use strict equality in date comparison.
- Remove generateWeekHeader method and directly insert week header string in run method.
- Remove trailing whitespace from weekly calendar date strings.

JunichiIto commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,78 @@
+
+  for (let day = 1; day <= lastDay; day++) {
+    monthDates += generateDayFormat(day);
+    if (isSaturday(year, month, day) || day === lastDay) {
+      monthDates = monthDates.trimEnd();
+      weeks.push(monthDates);
+      monthDates = "";
+    }
```
```suggestion
    if (isSaturday(year, month, day) || day === lastDay) {
      weeks.push(monthDates);
      monthDates = "";
    } else {
      monthDates += " ";
    }
```

一度追加したものを削除すると二度手間なので、こんな感じで「必要であれば追加する（不要なら何もしない）」というロジックの方がスマートかな〜と思います。

tuna24365 replied:
コメントありがとうございます！
ご提案のコードは処理も可読性もスマートでそちらの方が良いですね✨
あらためて自分のコードを省みると`trimEnd`案はその場しのぎ感がイけてないです💦

アドバイスの内容を受けて実際に修正しました。
末尾に空白を入れているのは`generateDayFormat`なので、このメソッドの余計な処理を無くして、`isSaturday`の条件分岐内で`「必要であれば追加する（不要なら何もしない）」`をしました😄

大規模な処理を実装する際は、処理量が増えるにつれて負荷が徐々に蓄積していくと思うので、こうした小さな点も気を付けたいと思います🙏

（`trimEnd`のままでは気持ちが良くないので、こちらもコミットに含めさせて頂こうと思います〜）

JunichiIto reviewed (approved):
1点コメントしましたが大きな問題ではないのでこれでOKです！🙆‍♂️

tuna24365 committed:
Optimize whitespace addition in calendar date generation.

JunichiIto commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,78 @@
+};
+
+const generateDayFormat = (day) => {
+  return day.toString().padStart(2, " ");
```
```suggestion
  return day.toString().padStart(2);
```

デフォルトが空白なので引数を省略してもいいかもしれません

https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/String/padStart

tuna24365 replied:
検収後にもかかわらずレビューありがとうございます😄
リファレンスを確認してみたらその旨の記載がありました📝

- padString
https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/String/padStart#padstring(省略可)
> 現在の文字列を延長するための文字列です。padString が targetLength より長い場合は、右からその長さ分が切り捨てられます。 `この引数の既定値は、Unicode の「空白」文字 (U+0020) です`。

引数を省略できるのは知らなかったです😲
（メソッドのリファレンスは使用する度にしっかり読み込みたいと思いました💡）
他にもpadStartを使用している箇所があるので、これらもご提案の内容に修正したいと思います〜

JunichiIto reviewed (approved):
修正内容もOKです〜 🙆‍♂️

tuna24365 committed:
Use default spacing in padStart method.
